File: backend/src/main/java/tech/wintrich/networkapi/presentation/controller/NetworkController.java
package tech.wintrich.networkapi.presentation.controller;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.constraints.NotBlank;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import tech.wintrich.networkapi.application.service.ConnectivityService;
import tech.wintrich.networkapi.application.service.DnsService;
import tech.wintrich.networkapi.application.service.HttpAnalysisService;
import tech.wintrich.networkapi.application.service.TlsService;
import tech.wintrich.networkapi.infrastructure.security.UrlValidator;
import tech.wintrich.networkapi.presentation.dto.DnsResponse;
import tech.wintrich.networkapi.presentation.dto.HttpAnalysisResponse;
import tech.wintrich.networkapi.presentation.dto.PingResponse;
import tech.wintrich.networkapi.presentation.dto.TlsInfoResponse;

/**
 * REST Controller for Network Intelligence API endpoints
 */
@Slf4j
@RestController
@RequestMapping("/api/network")
@RequiredArgsConstructor
@Validated
@Tag(name = "Network Intelligence", description = "API for analyzing connectivity, DNS, HTTP, and TLS")
@CrossOrigin(origins = "*")
public class NetworkController {

    private final ConnectivityService connectivityService;
    private final DnsService dnsService;
    private final HttpAnalysisService httpAnalysisService;
    private final TlsService tlsService;
    private final UrlValidator urlValidator;

    /**
     * Ping endpoint - checks connectivity and measures latency
     */
    @GetMapping("/ping")
    @Operation(
        summary = "Check connectivity and latency",
        description = "Performs a ping check to measure host reachability and latency"
    )
    public ResponseEntity<PingResponse> ping(
            @Parameter(description = "Target host or domain", example = "example.com")
            @RequestParam @NotBlank String host) {
        
        log.info("Ping request for host: {}", host);
        
        // Validate host
        urlValidator.validateHost(host);
        
        PingResponse response = connectivityService.ping(host);
        return ResponseEntity.ok(response);
    }

    /**
     * DNS lookup endpoint - retrieves DNS records
     */
    @GetMapping("/dns")
    @Operation(
        summary = "DNS intelligence lookup",
        description = "Retrieves comprehensive DNS records including A, AAAA, MX, NS, and TXT records"
    )
    public ResponseEntity<DnsResponse> dns(
            @Parameter(description = "Target domain", example = "example.com")
            @RequestParam @NotBlank String domain) {
        
        log.info("DNS lookup request for domain: {}", domain);
        
        // Validate domain
        urlValidator.validateHost(domain);
        
        DnsResponse response = dnsService.lookup(domain);
        return ResponseEntity.ok(response);
    }

    /**
     * HTTP analysis endpoint - analyzes HTTP response characteristics
     */
    @GetMapping("/http-analysis")
    @Operation(
        summary = "HTTP analysis and performance testing",
        description = "Analyzes HTTP response including status, headers, response time, and redirect chain"
    )
    public ResponseEntity<HttpAnalysisResponse> httpAnalysis(
            @Parameter(description = "Target URL", example = "https://example.com")
            @RequestParam @NotBlank String url) {
        
        log.info("HTTP analysis request for URL: {}", url);
        
        // Validate URL and check for SSRF
        urlValidator.validateUrl(url);
        
        HttpAnalysisResponse response = httpAnalysisService.analyze(url);
        return ResponseEntity.ok(response);
    }

    /**
     * TLS info endpoint - inspects SSL/TLS certificate
     */
    @GetMapping("/tls-info")
    @Operation(
        summary = "TLS/SSL certificate inspection",
        description = "Retrieves TLS protocol information and certificate details including issuer, expiration, and SANs"
    )
    public ResponseEntity<TlsInfoResponse> tlsInfo(
            @Parameter(description = "Target host", example = "example.com")
            @RequestParam @NotBlank String host) {
        
        log.info("TLS info request for host: {}", host);
        
        // Validate host
        urlValidator.validateHost(host);
        
        TlsInfoResponse response = tlsService.inspect(host);
        return ResponseEntity.ok(response);
    }
}

File: web-api/Controllers/NetworkController.cs
using Microsoft.AspNetCore.Mvc;
using Wintrich.NetworkApi.DTOs;
using Wintrich.NetworkApi.Infrastructure.Security;
using Wintrich.NetworkApi.Services;

namespace Wintrich.NetworkApi.Controllers;

/// <summary>
/// Network Intelligence REST controller.
/// Exposes the same four endpoints as the Java NetworkController:
///   GET /api/network/ping?host=
///   GET /api/network/dns?domain=
///   GET /api/network/http-analysis?url=
///   GET /api/network/tls-info?host=
/// </summary>
[ApiController]
[Route("api/network")]
[Produces("application/json")]
public sealed class NetworkController(
    ConnectivityService connectivityService,
    DnsService dnsService,
    HttpAnalysisService httpAnalysisService,
    TlsService tlsService,
    SecurityHeadersService securityHeadersService,
    UrlValidator urlValidator,
    ILogger<NetworkController> logger) : ControllerBase
{
    /// <summary>Check connectivity and measure ICMP latency for a host.</summary>
    [HttpGet("ping")]
    [ProducesResponseType<PingResponse>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> Ping([FromQuery] string? host)
    {
        if (string.IsNullOrWhiteSpace(host))
            return BadRequest(new { error = "host query parameter is required" });

        logger.LogInformation("Ping request for host: {Host}", host);
        urlValidator.ValidateHost(host.Trim()); // throws NetworkSecurityException → 403

        var response = await connectivityService.PingAsync(host.Trim());
        return Ok(response);
    }

    /// <summary>Retrieve A, AAAA, MX, NS and TXT DNS records for a domain.</summary>
    [HttpGet("dns")]
    [ProducesResponseType<DnsResponse>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> Dns([FromQuery] string? domain)
    {
        if (string.IsNullOrWhiteSpace(domain))
            return BadRequest(new { error = "domain query parameter is required" });

        logger.LogInformation("DNS lookup request for domain: {Domain}", domain);
        urlValidator.ValidateHost(domain.Trim());

        var response = await dnsService.LookupAsync(domain.Trim());
        return Ok(response);
    }

    /// <summary>Analyse HTTP response: status, headers, timings, redirect chain.</summary>
    [HttpGet("http-analysis")]
    [ProducesResponseType<HttpAnalysisResponse>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> HttpAnalysis([FromQuery] string? url)
    {
        if (string.IsNullOrWhiteSpace(url))
            return BadRequest(new { error = "url query parameter is required" });

        logger.LogInformation("HTTP analysis request for URL: {Url}", url);
        urlValidator.ValidateUrl(url.Trim());

        var response = await httpAnalysisService.AnalyzeAsync(url.Trim());
        return Ok(response);
    }

    /// <summary>Inspect the TLS certificate and connection details for a host on port 443.</summary>
    [HttpGet("tls-info")]
    [ProducesResponseType<TlsInfoResponse>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> TlsInfo([FromQuery] string? host)
    {
        if (string.IsNullOrWhiteSpace(host))
            return BadRequest(new { error = "host query parameter is required" });

        logger.LogInformation("TLS info request for host: {Host}", host);
        urlValidator.ValidateHost(host.Trim());

        var response = await tlsService.InspectAsync(host.Trim());
        return Ok(response);
    }

    /// <summary>Audit a URL for the presence of common HTTP security headers and assign a grade.</summary>
    [HttpGet("security-headers")]
    [ProducesResponseType<SecurityHeadersResponse>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status403Forbidden)]
    public async Task<IActionResult> SecurityHeaders([FromQuery] string? url)
    {
        if (string.IsNullOrWhiteSpace(url))
            return BadRequest(new { error = "url query parameter is required" });

        logger.LogInformation("Security headers audit request for URL: {Url}", url);
        urlValidator.ValidateUrl(url.Trim());

        var response = await securityHeadersService.AuditAsync(url.Trim());
        return Ok(response);
    }
}
